### ISSUE TITLE
Raiden Images contain outdated versions for `nightly`/`develop` tags.

### ISSUE THREAD

**Issue by nlsdfnbch** [raiden-network/raiden#5295]
Pulling the `nightly`/`develop` tag from `docker.io/raidennetwork/raiden` yields an outdated raiden version.

Logs
-------
```bash
(scenario-player) X280 /home/nls/devel/scenario-player/docker$ docker pull raidennetwork/raiden:nightly 
nightly: Pulling from raidennetwork/raiden
8d691f585fa8: Pull complete 
49bdb6f85638: Pull complete 
c0aff66c63e8: Pull complete 
ffec54d71584: Pull complete 
2189d0f348ac: Pull complete 
4becfb552b6c: Pull complete 
Digest: sha256:60bcdf01fa83b23b56719757015271fb1af4ac0be822fe6243ff23fc08c06766
Status: Downloaded newer image for raidennetwork/raiden:nightly
docker.io/raidennetwork/raiden:nightly
(scenario-player) X280 /home/nls/devel/scenario-player/docker$ docker run raidennetwork/raiden:nightly version  

{
  "raiden": "0.100.5a1.dev793+gb2d2674c4",
  "raiden_db_version": 23,
  "python_implementation": "CPython",
  "python_version": "3.7.5",
  "system": "Linux 64bit 5.0.0-35-generic",
  "architecture": "x86_64",
  "distribution": "source"
}
```

```bash
(scenario-player) X280 /home/nls/devel/scenario-player/docker$ docker run raidennetwork/raiden:develop version

Unable to find image 'raidennetwork/raiden:develop' locally
develop: Pulling from raidennetwork/raiden
8d691f585fa8: Already exists 
49bdb6f85638: Already exists 
78f19b867005: Pull complete 
7ebf582d4fcd: Pull complete 
9f71fc20b9f9: Pull complete 
dc9c3d84aa5f: Pull complete 
Digest: sha256:81873dd0ec2999311e0029473c6b69c7b6fe31d7038e03e8165bf58676caeebe
Status: Downloaded newer image for raidennetwork/raiden:develop
{
  "raiden": "0.100.5a1.dev795+geb173064c",
  "raiden_db_version": 23,
  "python_implementation": "CPython",
  "python_version": "3.7.5",
  "system": "Linux 64bit 5.0.0-35-generic",
  "architecture": "x86_64",
  "distribution": "source"
}
```

These versions should be built from the latest commit on the `develop` branch of the github repository, which is [currently](https://github.com/raiden-network/raiden/blob/eb173064c8acd81e2b1b4999e4656bb8cd8a154b/setup.py#L60) at `v0.200.0rc1` and some commits.

**Comment by nlsdfnbch** [comment]
After some RC pingpong with @kelsos and @LefterisJP , it appears like the Dockerfile does not pull the latest tags for the repository, as the commit sha seems to be pointing to the latest `develop` commit, making this bug a cosmetical one.

The reason the tags are required is that the versioning during the `make install` command depend on the git tags, instead of a version constant.

I'm unassigning mysefl from this, as I need to work on the SP instance.

**Comment by LefterisJP** [comment]
I will have a look. It may not be from the tags.

**Comment by LefterisJP** [comment]
All right figured it out. It's the tags but also the commit we tagged as `v0.200.0-rc1`.

We currently use [setuptools_scm](https://pypi.org/project/setuptools-scm/) in order to determine the current raiden version. What is returned from `raiden version` is the same as `python setup.py --version`.
The rules are the following (taken from above link):

![2019-11-15-150647_791x607_scrot](https://user-images.githubusercontent.com/1658405/68949095-9813cc80-07b9-11ea-95b8-0aeccae6511d.png)

Locally even with `raiden fetch upstream --tags` doing `python setup.py --version` returns `0.100.5a1.dev795+geb173064c`.

even though the latest tag should be https://github.com/raiden-network/raiden/releases/tag/v0.200.0-rc1

Turns out the reason is simple. The `v0.200.0-rc1` tag is not reachable from develop. 

```
git merge-base --is-ancestor v0.200.0-rc1 HEAD; echo $?                                                                                                                                                                        ST 10   develop 
1
```

That means master was not properly merged in develop after the `v0.200.0-rc1` release.

**Comment by LefterisJP** [comment]
Merging master into develop via a PR does not seem to work. The only way is to do the merge locally and push. But we are blocked by the branch protection.

We probably need to:

1. Remove the branch protection, push the “merging master to develop” develop and then reinstate it. This seems to be the only way to fix the wrong versions at the moment.

2. Additionally we will need to rethink this process as the way we currently make releases requires merging master back to develop which will need to be done manually and to remove the branch protection.

**Comment by Dominik1999** [comment]
Is this blocking our scenario runs? if so, is there a quickfix before the weekend?

**Comment by LefterisJP** [comment]
The quick fix is for me to do (1) but I want feedback on it. It's a bit heavy handed.

But I don't see why this would block the scenario runs since we have had this problem since the release was made. It's not new. I just realized it today thanks to this issue.

**Comment by LefterisJP** [comment]
Okay so after thinking on this the way we make releases from develop is currently broken. At the moment when we make a normal release we merge develop into master and then tag the merge commit in master.

For nightly releases we make a release from develop. That means that it is always going to be missing the last tag as seen in this issue.

The suggested solution would be to also tag develop before merging into master with a similar tag but a suffix indicating nightly.

For example:

master -> v0.200.0-rc1
develop -> v0.200.0-dev1

**Comment by ulope** [comment]
Doesn't look like master was merged back into develop:

![Screenshot 2019-11-20 at 10 51 57](https://user-images.githubusercontent.com/55078/69228591-1ee2f380-0b84-11ea-92c7-ac4e22d10036.png)

**Comment by LefterisJP** [comment]
> One downside to this: -devX sorts earlier than -rcX according to PEP440 rules.

Then we need to find a suffix combination that works. Perhaps reverse rc and dev in our tags?

> It doesn't look like master was actually merged back into develop:

@ulope I think a PR to do so was made but closed: https://github.com/raiden-network/raiden/pull/5037

5 days ago I tried to do the same thing but it was a no-op: https://github.com/raiden-network/raiden/pull/5300
 
It seems merging master to develop via a PR does not work. It would have to be done manually and then pushed to upstream and for that to happen all repository checks would need to be disabled as the upstream develop branch is protected.

**Comment by ulope** [comment]
@karlb Do you remember if you used "merge" or "rebase" when merging that PR?

Hm wait, didn't we even disable merge commits in the settings?

**Comment by karlb** [comment]
> @karlb Do you remember if you used "merge" or "rebase" when merging that PR?

I used the default action. That should have been rebase, but I don't explicitly remember that.

**Comment by ulope** [comment]
Yes we did disable merge commits: ![](https://ds.ulo.pe/QHjNc0ECd2.png)
That's the problem. Rebase changes the commit ids therefore tags don't "come across".

**Comment by LefterisJP** [comment]
It must have been a rebase. All other options are disabled from the settings. That may explain it then.

**Comment by LefterisJP** [comment]
So the solution here would be to allow merge commits and ask for discipline from the team to choose the correct option? And then use a proper merge for merging master back into develop?

**Comment by ulope** [comment]
@konradkonrad and I just went through something very similar in the raiden-service-bundle repo. 

Seems like even a merge commit is taking the tags along. Needs some more looking into.

**Comment by LefterisJP** [comment]
Perhaps the solution I suggested [above](https://github.com/raiden-network/raiden/issues/5295#issuecomment-555446320) but with more well thought out suffixes?

**Comment by ulope** [comment]
I would consider that a last resort.
A tag should uniquely identify a specific version of the code. The branch shouldn't matter.

**Comment by ulope** [comment]
Thinking more about this: Were the tags actually pushed to github? 
By default that is not the case and would explain why the merge commit did not work.

Tags are only pushed if:
- `git push --tags` (explicitly push all local tags)
- `git push `[--follow-tags](https://git-scm.com/docs/git-push#Documentation/git-push.txt---follow-tags) (push all reachable tags, e.g. when pushing a feature branch)
- The `push.followTags` config option has been enabled (automates the behaviour from above)
